pkg/webhook/mutation.go

```diff
@@ -157,9 +165,57 @@ func (h *mutationHandler) Handle(ctx context.Context, req admission.Request) adm
 }
 
 func (h *mutationHandler) mutateRequest(ctx context.Context, req admission.Request) (admission.Response, error) {
-	// TODO: place mutation logic here
-	mutatedObject := req.Object
-	resp := admission.PatchResponseFromRaw(req.Object.Raw, mutatedObject.Raw)
+
+	ns := &corev1.Namespace{}
+
+	// if the object being mutated is a namespace itself, we use it as namespace
+	if req.Kind.Kind == namespaceKind && req.Kind.Group == "" {
+		req.Namespace = ""
+		obj, _, err := deserializer.Decode(req.Object.Raw, nil, &corev1.Namespace{})
+		if err != nil {
+			return admission.Errored(int32(http.StatusInternalServerError), err), nil
+		}
+		ok := false
+		ns, ok = obj.(*corev1.Namespace)
+		if !ok {
+			return admission.Errored(int32(http.StatusInternalServerError), errors.New("failed to cast namespace object")), nil
+		}
+	} else if req.AdmissionRequest.Namespace != "" {
+		if err := h.client.Get(ctx, types.NamespacedName{Name: req.AdmissionRequest.Namespace}, ns); err != nil {
+			if !k8serrors.IsNotFound(err) {
+				log.Error(err, "error retrieving namespace", "name", req.AdmissionRequest.Namespace)
+				return admission.Errored(int32(http.StatusInternalServerError), err), nil
+			}
+			// bypass cached client and ask api-server directly
+			err = h.reader.Get(ctx, types.NamespacedName{Name: req.AdmissionRequest.Namespace}, ns)
+			if err != nil {
+				log.Error(err, "error retrieving namespace from API server", "name", req.AdmissionRequest.Namespace)
+				return admission.Errored(int32(http.StatusInternalServerError), err), nil
+			}
+		}
+	}
+	obj := unstructured.Unstructured{}
+	err := obj.UnmarshalJSON(req.Object.Raw)
+	if err != nil {
+		log.Error(err, "failed to unmarshal", "object", string(req.Object.Raw))
+		return admission.Errored(int32(http.StatusInternalServerError), err), nil
```

Comment:
I'm wondering if the MWH failure policy also applies to 400 errors?

Not something to address in this PR, but if it does we can make sure we are returning the most semantically correct error code.